packages/agents/sequencer/src/lib/helpers/auctions.ts

```diff
@@ -63,3 +64,83 @@ export const getDestinationLocalAsset = async (
   const localAddress = destinationDomainAsset!.local;
   return localAddress;
 };
+
+/**
+ * Picks up the valid bids and groups the bids by round
+ * @param bids - The array of raw bids, mapping between `router` and `bid` for the specific transferId
+ * @param roundDepth - The maximum round which is allowed in the sequencer
+ * @returns roundId -> array of bids
+ */
+export const getBidsRoundMap = (bids: Record<string, Bid>, roundDepth: number): Record<number, Bid[]> => {
+  const availableBids: Record<number, Bid[]> = {};
+  // Sanity checks and pick up valid bid combinations
+  for (let roundIdx = 1; roundIdx <= roundDepth; roundIdx++) {
+    const filteredBids = Object.values(bids)
+      .filter((bid) => {
+        return Array.from(Object.keys(bid.signatures)).includes(roundIdx.toString());
+      })
+      .map((bid) => {
+        const signatures: Record<string, string> = {};
+        signatures[roundIdx.toString()] = bid.signatures[roundIdx].toString();
+        return { ...bid, signatures };
+      });
+
+    // We need 2 ^ (roundIdx - 1) of different bids at least for roundIdx
+    if (filteredBids.length >= getMinimumBidsCountForRound(roundIdx)) {
+      availableBids[roundIdx] = filteredBids;
+    }
+  }
+
+  return availableBids;
+};
+
+/**
+ * Calculates the minimum number of bids for the round
+ * round 1 -> amount / 1, round 2 -> amount / 2, round 3 -> amount / 4, round4 -> amount / 8
+ * @param round The round id to calculate the minimum number of bids for
+ * @returns - The number of bids
+ */
+export const getMinimumBidsCountForRound = (round: number): number => {
+  const { config } = getContext();
+  if (round < 1 || round > config.maxPathLength || Math.trunc(round) != round) {
+    throw new RoundInvalid({ round, maxPathLength: config.maxPathLength });
+  }
+  return Math.pow(2, round - 1);
+};
+
+/**
+ * Generate all combinations of an array.
+ * @param sources - Array of input elements.
+ * @param length - Desired length of combinations.
+ * @return - Array of combination arrays.
+ */
+export const generateCombinations = (sources: any[], length: number): any[] => {
```

Comment:
total nitpick but - this isn't generating arrays, it's generating **sets**. If I understand it correctly, this function is really more like `getAllSubsets`. I'd prefer a name like that.

difference ofc being that, for arrays, order matters, for sets, it doesn't

this doesn't need to return a ts Set object, should still just return an array, but yeah verbage here is important lol

Reply:
it actually returns an array of combinations, not sets

Reply:
I know the type is an array, but mathematically it's a subset

every length-2 subset of A, B, C:
A, B
B, C
A, C

every length-2 sub-array of A, B, C:
A, B
B, A
B, C
C, B
A, C
C, A

not too important, just felt like it was worth noting in a comment at least, idk